### ISSUE TITLE
Deleting a file in a big project takes over 3 minutes and then times out

### ISSUE THREAD

**Issue by Dzarg** [SRombauts/UE4PlasticPlugin#70]
Hi again,

Thank you again for solving the problem we had with history and diffs. We've been happy with how things have been regarding those.

Now another pretty normal use case we have is deleting a file. Deleting a single file seems to trigger a status check for 57k files (files inside Content).

First PlasticSourceControlUtils::RunStatus takes about 5 seconds (long time, but not too long):
![image](https://user-images.githubusercontent.com/94673/70803269-5bd58b00-1dbc-11ea-93c2-6028f7438a38.png)

Then PlasticSourceControlUtils::RunUpdateStatus is called asking fileinfo for 57k files. That in turn causes PlasticSourceControlUtils::_RunCommandInternal to try to read something from the cm process, but won't get anything in 3 minutes and times out.

Would there be anything that can be done to make this better?

Thanks in advance!

**Comment by SRombauts** [comment]
Hello, thanks for the report!

I'll have to check that, I am not sure what is making this, it would be interesting to see if this behavior is triggered by the Engine itself, or by my plugin.

**Comment by Dzarg** [comment]
Hi,

I checked this more and my test case was as follows:
1) Create a new blueprint with actor as parent
2) Delete it

This one also does it:
1) Create a new blueprint with actor as parent
2) Rename it

Probably move as well, but I didn't test it yet. Anything that causes a delete to happen basically. And I assume it happens with other types of assets as well.

But seems that for something that was not just created it works fine. 

If I'm understanding this correctly, when `ObjectTools::CleanupAfterSuccessfulDelete` can't find a saved package for the deleted/renamed/moved file (I'm assuming for a file that has never been saved yet), then it still calls `SourceControlProvider.GetState(PackageFilesToDelete, PackageSCCStates, EStateCacheUsage::ForceUpdate);` This then seems to trigger a special case in `FPlasticUpdateStatusWorker::Execute` when there are no infiles, it updates the whole content dir.

To me this feels a problem already from the engine side. Why is calling the function without files? Though I guess they expect the call without a files to go ~instantly.

The comments in `ISourceControlProvider::GetState` don't mention any special case for when no files is given:
```
	/**
	 * Get the state of each of the passed-in files. State may be cached for faster queries. Note states can be NULL!
	 * @param	InFiles				The files to retrieve state for.
	 * @param	OutState			The states of the files. This will be empty if the operation fails. Note states can be NULL!
	 * @param	InStateCacheUsage	Whether to use the state cache or to force a (synchronous) state retrieval.
	 * @return the result of the operation.
	 */
```

So I would assume that in this case the plastic plugin shouldn't do anything. What do you think?

**Comment by Dzarg** [comment]
I did some more testing and everything I try seem to be working well if I take out this block of code:
```
	else
	{
		// no path provided: only update the status of assets in Content/ directory and also Config files
		TArray<FString> ProjectDirs;
		ProjectDirs.Add(FPaths::ConvertRelativePathToFull(FPaths::ProjectContentDir()));
		ProjectDirs.Add(FPaths::ConvertRelativePathToFull(FPaths::ProjectConfigDir()));
		InCommand.bCommandSuccessful = PlasticSourceControlUtils::RunUpdateStatus(ProjectDirs, false, InCommand.Concurrency, InCommand.ErrorMessages, States, InCommand.ChangesetNumber, InCommand.BranchName);
	}
```

If this block is taken out, will it cause problems in some situations?

**Comment by SRombauts** [comment]
Hey, thanks for taking the time. I actually new where it would lead, just didn't understand why it doesn't happen in my usual testings.
So this special case I added is only need for either the initial creation of the workspace and then at each Editor Startup if you don't disable the option to update the whole workspace state

Whole workspace update is useful only to populate the cache, to get proper result with content browser source control filters or when right clicking on a folder
I also added on the Perforce plugin, it is reasonable only up to a certain project size

Issue with this current reasoning: How is it that the initial call take 5s to refresh the whole repo, while this one seems to be listing all files of the repo? I might be confusing different use cases.

I'll see if I can take some time to look at it during the weekend

**Comment by Dzarg** [comment]
Thank you for the info.

I made a slight mistake in my first post describing the situation.

When deleting a file then `PlasticSourceControlUtils::RunUpdateStatus` is called and inside it there's first call to `PlasticSourceControlUtils::RunStatus` that takes about 5s and continuing in the same function next `PlasticSourceControlUtils::RunFileinfo` is called and this causes the 3 min timeout to happen.

Here's a better zoomed/panned image of the same profile:
![image](https://user-images.githubusercontent.com/94673/70844979-e74d2b80-1e51-11ea-88ba-41e3e04c0e81.png)
The green area at the top is when the thread is using cpu time and from here we can see that during the `RunFileInfo` (which still continues a long way to the right) it's not using almost any cpu. Sleeping 98% of the time according to Superluminal.

I haven't checked how it's done but I don't think our project does source control refresh at editor start and also the filters don't work for us. So this explains it as well.

For now I can just locally take out the block I mentioned above as I don't think the operation ever completes anyway in our project. So this is not a high priority issue anymore for us as we now should have a working workaround. Please have a relaxing weekend :)

**Comment by reveracorp** [comment]
Same issue, sometimes need to wait minutes for Updating . . . to happen and time out. Can be painful because it stops me in my tracks when it happens.

**Comment by SRombauts** [comment]
Ok, thanks for the head up @reveracorp, do you have any repro case?

I was unable to reproduce, perhaps my project is not big enough, or my Plastic SCM setup is faster and don't time out?

I would love some more insights on this, like logs for instance.

**Comment by SRombauts** [comment]
A number of performance issues have been fixed [in recent releases (esp. 1.4.12 and 1.4.13)](https://github.com/SRombauts/UE4PlasticPlugin/releases)

I am quite confident it is now fixed with 1.4.13
